src/rgw/rgw_bucket.cc

```diff
@@ -2181,6 +2181,141 @@ class RGWBucketMetadataHandler : public RGWMetadataHandler {
     return 0;
   }
 
+  int remove_on_archive_zone(RGWRados *store, string& metadata_key) override {
+
+    Formatter *f = new JSONFormatter(false);
+    int ret = store->meta_mgr->get(metadata_key, f);
+    if (ret < 0) {
+        lderr(store->ctx()) << "can't get key: " << metadata_key << dendl;
+        delete f;
+        return ret;
+    }
+
+    string type;
+    string entry;
+    store->meta_mgr->parse_metadata_key(metadata_key, type, entry);
+    string tenant_name, bucket_name;
+    parse_bucket(entry, &tenant_name, &bucket_name);
+
+    RGWMetadataObject *obj;
+    ret = get(store, entry, &obj);
+    if (ret < 0) {
+        delete f;
+        return ret;
+    }
+
+    RGWBucketEntryPoint be;
+    RGWObjectCtx obj_ctx(store);
+    RGWObjVersionTracker objv_tracker;
+    objv_tracker.read_version = obj->get_version();
+    delete obj;
+    ret = store->get_bucket_entrypoint_info(obj_ctx, tenant_name, bucket_name, be, &objv_tracker, NULL, NULL);
+    if (ret < 0) {
+        delete f;
+        return ret;
+    }
+
+    ret = rgw_unlink_bucket(store, be.owner, tenant_name, bucket_name, false);
+    if (ret < 0) {
+        lderr(store->ctx()) << "could not unlink bucket=" << entry << " owner=" << be.owner << dendl;
+    }
+
+    ret = rgw_bucket_delete_bucket_obj(store, tenant_name, bucket_name, objv_tracker);
+    if (ret < 0) {
+        lderr(store->ctx()) << "could not delete bucket=" << entry << dendl;
+    }
+
+    string new_metadata_key = metadata_key + "-deleted";
+
+    string new_entry, new_bucket_name;
+    store->meta_mgr->parse_metadata_key(new_metadata_key, type, new_entry);
+    parse_bucket(new_entry, &tenant_name, &new_bucket_name);
+
+    bufferlist bl;
+    f->flush(bl);
+    delete f;
+
+    string bl_str = bl.to_str();
+
+    // fix key
+    size_t found = bl_str.find(entry);
```

Comment:
@jmunhoz this whole find and replace here is not necessary and is dangerous anyway as it relies on the fact that the json representation can be easily patched and there will never be any duplicates. I think the way to do it is by parsing the original entrypoint, and then modify the structure directly.

Reply:
We need the functionality to update the omap keys and so on. Consider this example without the find&replace code to illustrate the list-buckets case by user testuser1:

$ aws --endpoint=http://rgw1:8000 s3api create-bucket --bucket my-bucket
$ aws --endpoint=http://rgw1:8000 s3api delete-bucket --bucket my-bucket
$ aws --endpoint=http://rgw2:8001 s3api list-buckets
{
    "Owner": {
        "DisplayName": "Test User", 
        "ID": "testuser1"
    }, 
    "Buckets": [
        {
            "CreationDate": "2018-10-03T13:35:19.406Z", 
            "Name": "my-bucket"
        }
    ]
}

$ ./bin/rados -p us-east-2.rgw.meta -N users.uid listomapkeys testuser1.buckets
my-bucket
$ ./bin/rados -p us-east-2.rgw.meta -N root ls | sort
.bucket.meta.my-bucket-deleted-4a292cbbedba1047528464307bd9bc26:8c829b5a-4f7e-4843-a454-78b32c0097e8.4226.1
my-bucket-deleted-4a292cbbedba1047528464307bd9bc26

The find&replace looks a very fast but also hardwired way to replace the two occurrences for the bucket name. As you indicate, a better way is by parsing the original entrypoint, and then modify the structure directly.